### ISSUE TITLE
Bring video props to root, and update Typescript definition to reflect this

### ISSUE THREAD

**Issue by hastarin** [devcshort/react-hls#4]
I'm new to React, Typescript, etc but have managed to use this component with Typescript by creating a custom [type definition file](https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html).

```typescript
declare module 'react-hls-player' {

	export class ReactHls extends React.Component<DetailedHTMLProps<VideoHTMLAttributes<HTMLVideoElement>, HTMLVideoElement>> {
		url: string;
		hlsConfig?: Partial<Hlsjs.Config>;
		controls?: boolean;
	}

	export default ReactHls;
}
```

I'm unsure if it's correct but it got the basics working for me.   Hopefully this helps someone else until someone more experienced than myself could perhaps improve the definition and submit a pull request.

**Comment by devcshort** [comment]
Thanks for the feedback @hastarin. My knowledge in Typescript is fairly limited, but I agree that proper Typescript definitions will definitely make a great addition. I originally forked this project as it looked like it was abandoned and was one I needed. My goal was to fix issues that I was running in to, while also maintaining the package and adding new features. I am currently looking in to the possibility of a full rewrite and releasing as version 1.

I will do more research in to this and as time allows will add Typescript compatibility. I'll leave this issue open for now so others can see a possible solution until then.

**Comment by devcshort** [comment]
@hastarin I've released a V1 of this package. It now includes a typescript definition file by default for all that intellisense goodness! API should be the same, this was just a rewrite a long with some improvements.

7b8979cafb8242b9fd4794fff498b7969dd92b55

**Comment by hastarin** [comment]
Thanks @devcshort I will have to make time to update my little test site I am using this in.

At a glance though something closer to the definitions I used would have worked a lot better?

You're actually extending existing controls and I would have thought it's easier to use their definitions than redefine them.

I'll have to try it to see, but would think that with what you've gone with some of the magic of Typescript, such as Intellisense for various properties, will be unavailable as you've defined them as objects.

Again, I'm no expert, my language of choice is C# which is why I've been using Typescript for a little side project in the first place.

**Comment by devcshort** [comment]
> Thanks @devcshort I will have to make time to update my little test site I am using this in.
> 
> At a glance though something closer to the definitions I used would have worked a lot better?
> 
> You're actually extending existing controls and I would have thought it's easier to use their definitions than redefine them.
> 
> I'll have to try it to see, but would think that with what you've gone with some of the magic of Typescript, such as Intellisense for various properties, will be unavailable as you've defined them as objects.
> 
> Again, I'm no expert, my language of choice is C# which is why I've been using Typescript for a little side project in the first place.

You’re completely right. I shouldn’t have extended it the way I did. My knowledge of Typescript is minimal and I was fumbling with it earlier. I was too focused on some of the extras that I completely missed the proper typings for some of the things I did as objects. I’ll push out a fix as soon as possible to have better typings. Thanks for the feedback!

**Comment by devcshort** [comment]
@hastarin I tried to update my Typescript definition so that it could extend the HTML Video Element, but I started running in to issues from that. It would tell me that all of the attributes are considered required (width, height, autoplay, etc). Did you run in to this at all? I don't want to force this requirement if I'm to have the hls video player be an extension of that. I also ran in to issues when trying to overwrite the style attribute to use React's style interface.

I'm going to reopen this in case anyone else may be able to help as well. I want to try and get all the Typescript definitions working properly.

**Comment by hastarin** [comment]
> @hastarin I tried to update my Typescript definition so that it could extend the HTML Video Element, but I started running in to issues from that. It would tell me that all of the attributes are considered required (width, height, autoplay, etc). Did you run in to this at all? I don't want to force this requirement if I'm to have the hls video player be an extension of that. I also ran in to issues when trying to overwrite the style attribute to use React's style interface.

No I did not.  The only required attribute in the definition I used is url.  Make sure if you're doing something like the following:
```hlsConfig?: Partial<Hlsjs.Config>;```

You need to use Partial so it doesn't require all attributes from the referenced type.

**Comment by devcshort** [comment]
> No I did not. The only required attribute in the definition I used is url. Make sure if you're doing something like the following:
> `hlsConfig?: Partial<Hlsjs.Config>;`
> 
> You need to use Partial so it doesn't require all attributes from the referenced type.

I had tried the Partial based on your example, which worked for the Hlsjs Config, but I was also trying to have the base extend from the html video component which is where I was running in to issues. Essentially I want all html video component props to be accessible as root props rather than being a separate object. As soon as I extended HTMLVideoElement I received errors that all video element props are required. Maybe I should replicate your example completely rather than just making an interface. 🤔

Or is it possible to extend Partial\<HTMLVideoElement\>? I'll have to look in to it and see.

Thanks!

**Comment by devcshort** [comment]
Just released version 1.1.0. I updated the Typescript declaration to use your example. I made one small change to how the types extend as I was getting class based component methods showing up in intellisense for props. It seems to be working as it should now. Let me know if you think there are any other changes I should make to it.

Thanks again!
